sql/src/main/java/org/apache/druid/sql/calcite/rel/DruidQuerySignature.java

```diff
@@ -0,0 +1,167 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.sql.calcite.rel;
+
+import com.google.common.collect.ImmutableMap;
+import org.apache.calcite.sql.type.SqlTypeName;
+import org.apache.druid.segment.VirtualColumn;
+import org.apache.druid.sql.calcite.expression.DruidExpression;
+import org.apache.druid.sql.calcite.planner.Calcites;
+import org.apache.druid.sql.calcite.planner.PlannerContext;
+import org.apache.druid.sql.calcite.table.RowSignature;
+
+import javax.annotation.Nullable;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.TreeSet;
+
+/**
+ * Wraps a {@link RowSignature} and provides facilities to re-use {@link VirtualColumn} definitions for dimensions,
+ * filters, and filtered aggregators while constructing a {@link DruidQuery}
+ */
+public class DruidQuerySignature
+{
+  private final RowSignature rowSignature;
+
+  private final Map<String, VirtualColumn> virtualColumnsByExpression;
+  private final Map<String, VirtualColumn> virtualColumnsByName;
+  private final String virtualColumnPrefix;
+  private int virtualColumnCounter;
+
+  private final boolean isImmutable;
+
+  public DruidQuerySignature(RowSignature rowSignature)
```

Comment:
Hmm, I'm not actually sure it should legitimately be `@Nullable`, currently the only usage of `null` is a test where it doesn't matter. Maybe I should modify that test instead to pass in a `RowSignature` instead?

Reply:
Hmm, I thought it can be null if the query is nested and the inner query doesn't have the output row signature (I don't know what it means though..). See `DruidOuterQueryRel.toDruidQuery()` and `DruidQuery`.

Reply:
Traced through and I think this can never be `null`; I've removed `@Nullable` from the field in `DruidQuery`.

Reply:
You're right. Thanks.